### PR TITLE
Handle anonymous class attribute

### PR DESCRIPTION
**Pull request by romm**
Hi, first of all thanks for your work, I didn't know about this library and it could be of great use for a test suite in a project of mine.

I ran into the following issue: when an anonymous class uses a PHP 8 attribute, Patchwork crashes.

I'm adding a failing test to this PR, I currently have no idea where to look to fix this, I'd be glad to help if someone gives me a hint. 🙂 

Have a good day!

### CODE REVIEW

**Comment by anomiex** [comment]
> I currently have no idea where to look to fix this, I'd be glad to help if someone gives me a hint. 🙂

Looks like the code in `injectCodeAfterClassDefinitions` skips over whitespace and comments (see `Source::junk()`), but to handle the `new class` case it also needs to skip over attributes.
https://github.com/antecedent/patchwork/blob/27e8ab912e7e433ef0d5a8aa4c19e6a56f0839d0/src/CodeManipulation/Actions/Generic.php#L140-L143

Also, I find that `php -r 'require "./Patchwork.php"; echo Patchwork\CodeManipulation\transformString( file_get_contents( "tests/includes/AnonymousClassAttribute.php" ) );'` can show what Patchwork is turning the code in your test into, which may help in working on it.

**Comment by jrfnl** [comment]
Re: the tests - probably a good idea to add some more variations, like multi-line attributes with array parameters + a class multiple attributes attached, as those cases will need to be handled correctly and are, based on experience elsewhere, the ones which will break any "simple" solution.

Another thing which should probably be taken into account (but solved separately) is PHP 8.3 readonly anonymous classes, where the `readonly` keyword between `new` and `class` could possibly trip Patchwork up ?

**Comment by romm** [comment]
@anomiex @jrfnl thanks for your feedback, I've just pushed some changes and will be waiting for further comments or approval. 🙂

File: src/CodeManipulation/Actions/Generic.php
Comment (antecedent) [reply src/CodeManipulation/Actions/Generic.php:140]: I think that all named classes have a T_STRING token (the name) following the `class` token. This, however, only comes from my intuition for the time being, and could be wrong. But if it is correct, then we could get around this issue by inspecting the token on the right side of the `class` token, not the left side.

Comment (jrfnl) [reply src/CodeManipulation/Actions/Generic.php]: > I think that all named classes have a T_STRING token (the name) following the `class` token. This, however, only comes from my intuition for the time being, and could be wrong. But if it is correct, then we could get around this issue by inspecting the token on the right side of the `class` token, not the left side.

Depends on how the tokenizer is run. Looks like it is run with the `TOKEN_PARSE` `$flag`, in which case _probably yes_, though results from running the tokenizer with that flag vary depending on the PHP version (then again, that is the case for token streams anyway).

In PHPCS, to determine whether the `class` keyword is an anonymous class or a named class, we check if the next non-empty token is one of the following, if it is, the `class` keyword is marked as an anonymous class:
* `(` (open parenthesis)
* `{` (open curly)
* `T_EXTENDS`
* `T_IMPLEMENTS

Comment (anomiex) [reply src/CodeManipulation/Actions/Generic.php]: Personally what I was thinking of doing was to `$pos = $s->skipBack(Source::junk(), $match)`, then if it's pointing at a `RIGHT_SQUARE` and `$s->match()` for that points to a `General\ATTRIBUTE`, `skipBack` from that `General\ATTRIBUTE` and repeat as necessary for more `RIGHT_SQUARE`s. Once it finds a non-bracket (or a `RIGHT_SQUARE` not matching with a `General\ATTRIBUTE`), then look for the `T_DOUBLE_COLON` or `T_NEW`.

Although I like @jrfnl 's idea too.

Comment (antecedent) [reply src/CodeManipulation/Actions/Generic.php]: I took @jrfnl 's suggestion. Both the last one, and the previous one about `readonly`. T_READONLY really seems to be the only token that could intervene between `new` and `class`, beside attributes: https://github.com/nikic/PHP-Parser/blob/master/grammar/php.y#L501-L515

**Comment by antecedent** [comment]
@romm, could you run this PR's branch with your intended code base, as an end-to-end test of sorts? Just to make sure that no other bugs show up.

**Comment by romm** [comment]
Hi @antecedent, just tried the patch on my code base and it works like a charm. I detected another issue regarding private constructors, but it seems to be a completely different problem, I'll investigate on it.

Thanks!

**Review by antecedent — APPROVED** [review]

**Pull request merged**
